### PR TITLE
[WIP] New pihole checkout feature

### PR DESCRIPTION
**Pull request by DL6ER**
**By submitting this pull request, I confirm the following (please check boxes, eg [X]) _Failure to fill the template will close your PR_:**

***Please submit all pull requests against the `development` branch. Failure to do so will delay or deny your request***

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] I have checked that [another pull request](https://github.com/pi-hole/pi-hole/pulls) for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**How familiar are you with the codebase?:** 10

---
Add a feature that allows checking out a different branch even from an initially shallow clone.

Available shortcut:
- `pihole checkout dev` -> check out to `development` + `devel`

_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

### CODE REVIEW

**Comment by Mcat12** [comment]
Can you move the code from `pihole` into the script?

File: advanced/Scripts/piholeCheckout.sh [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  r=$(( rows ))
+  c=$(( columns ))
+  # Unless the screen is tiny
+  r=$(( r < 20 ? 20 : r ))
+  c=$(( c < 70 ? 70 : c ))
+
+  local cnt=${#branches[@]}
+
+  for ((i=0;i<cnt;i++)); do
+    if [[ ${branches[i]} == "master" ]]; then
```
Comment (Mcat12) [inline advanced/Scripts/piholeCheckout.sh]: This should set the current branch as on, even if it's not master

File: advanced/Scripts/piholeCheckout.sh [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+  curdir="${PWD}"
+  cd "${directory}" || return 1
+  # Get reachable remote branches
+  git remote show origin | grep 'tracked' | sed 's/tracked//;s/ //g'
+  cd "${curdir}" || return 1
+  return
+}
+
+choice=""
```
Comment (dschaper) [inline advanced/Scripts/piholeCheckout.sh]: `choice` doesn't look like it's being used anywhere else in the script. May be vestigial.

File: advanced/Scripts/piholeCheckout.sh [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+select_branch() {
+  branches=($@)
+  # Divide by two so the dialogs take up half of the screen, which looks nice.
+  rows=$(echo $screen_size | awk '{print $1}')
```
Comment (dschaper) [inline advanced/Scripts/piholeCheckout.sh]: `$screen_size` isn't declared anywhere, need to `stty size` to initialize it.

File: advanced/Scripts/piholeCheckout.sh [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  for ((i=0;i<cnt;i++)); do
+    if [[ ${branches[i]} == "master" ]]; then
+      branches[i]="\"${branches[i]}\" \"\" on"
+    else
+      branches[i]="\"${branches[i]}\" \"\" off"
+    fi
+  done
+
+  # Display dialog
+  ChooseCmd=(whiptail --radiolist \"Select target branch\" ${r} ${c} ${#branches[@]})
```
Comment (dschaper) [inline advanced/Scripts/piholeCheckout.sh]: If the whiptail is cancelled out then the following `echo` to `checkout-choice` would be blank, and that can be handled in the caller script.

File: advanced/Scripts/piholeCheckout.sh [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+  r=$(( rows / 2 ))
+  c=$(( columns / 2 ))
+  # Unless the screen is tiny
+  r=$(( r < 20 ? 20 : r ))
+  c=$(( c < 70 ? 70 : c ))
+  if (whiptail --yesno "Note that changing the branch is a severe change of your Pi-hole system.
+This is not supported unless one of the developers explicitly asks you to do this!
+
+Have you read and understood this?" ${r} ${c} --title "----> WARNING <----") then
+    return 0
```
Comment (dschaper) [inline advanced/Scripts/piholeCheckout.sh]: Might be able to just call the whiptail and return `$?`, should be 0 - yes and 1 - no.

File: pihole [hunk 6]
```diff
@@ -218,6 +225,65 @@ tailFunc() {
...
+    checkout_branch "${PI_HOLE_FILES_DIR}" "development"
+    echo "::: Web interface"
+    checkout_branch "${ADMIN_INTERFACE_DIR}" "devel"
+    echo "::: done!"
+  else
+    # Have to user chosing the branch he wants
+    select_branch "${corebranches[@]}"
+    corebranch=$(cat /etc/pihole/checkout-choice)
+    select_branch "${webbranches[@]}"
+    webbranch=$(cat /etc/pihole/checkout-choice)
```
Comment (dschaper) [inline pihole]: Bash doesn't return non-numerics very well, but I'd check here for an empty file (which would mean a cancelled whiptail.)

**Comment by DL6ER** [comment]
Work will be continued in new pull request

**Pull request closed without merging**